<a href="https://colab.research.google.com/github/lauramazzu/hackbio-biocoding-internship/blob/main/**AgePredictionsModel**.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [151]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import RFE
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


from sklearn.ensemble import RandomForestRegressor

**Age Prediction in Humans**

Study the data here

There are 1000 CpG sites to be used to
determine the age of 108 individuals (Homo sapiens).

Using any regression model, predict the age of the patients (last columns).

In your feature selection step, reduce the number of features to select to 90, 80, 70, 60, 50, 40, 30, 20 and 10.

 How does the MSE behave with the reducing number of features?

Generate a plot of the number of features against the mse values.

In [152]:
#Loading the dataset
df_sites_link = "https://raw.githubusercontent.com/HackBio-Internship/2025_project_collection/refs/heads/main/Python/Dataset/methylation.data.csv"

df_sites = pd.read_csv(df_sites_link)
df_sites.head()
print(df_sites.head())
print(f"The dataset has {df_sites.shape[0]} rows and {df_sites.shape[1]} .")

   Unnamed: 0  cg22051636  cg25569462  cg03606258  cg22022041  cg00729708  \
0           1    0.028607    0.118231    0.354948   -0.195587   -0.328310   
1           2   -0.421796   -0.484327    0.358169    0.184083    0.058353   
2           3   -0.425853   -0.486174    0.017406   -0.183193    0.227653   
3           4   -0.449498    0.136772    0.366528   -0.169309    0.151502   
4           5    0.230452    0.232878    0.313735    0.187171   -0.371522   

   cg07747299  cg14159672  cg12437481  cg20945531  ...  cg25307081  \
0   -0.175936   -0.169272    0.166648    0.013308  ...   -0.043515   
1   -0.173480   -0.135551   -0.196084    0.024734  ...   -0.016707   
2    0.415791   -0.077909   -0.043989    0.103141  ...   -0.023028   
3   -0.191507   -0.125910    0.200092   -0.013143  ...   -0.004660   
4   -0.185925    0.564402   -0.209667   -0.318550  ...   -0.003939   

   cg02298570  cg03545635  cg07310951  cg02712878  cg04697562  cg20090467  \
0   -0.042265   -0.087791   -0.030244  

In [182]:
#change the first column from "unnamed" to "patient number" to make it easier to understand
df_sites.rename(columns={'Unnamed: 0': 'Patient Number'}, inplace=True)
print(df_sites.head())
df_sites.head()

   Patient Number  cg22051636  cg25569462  cg03606258  cg22022041  cg00729708  \
0               1    0.028607    0.118231    0.354948   -0.195587   -0.328310   
1               2   -0.421796   -0.484327    0.358169    0.184083    0.058353   
2               3   -0.425853   -0.486174    0.017406   -0.183193    0.227653   
3               4   -0.449498    0.136772    0.366528   -0.169309    0.151502   
4               5    0.230452    0.232878    0.313735    0.187171   -0.371522   

   cg07747299  cg14159672  cg12437481  cg20945531  ...  cg25307081  \
0   -0.175936   -0.169272    0.166648    0.013308  ...   -0.043515   
1   -0.173480   -0.135551   -0.196084    0.024734  ...   -0.016707   
2    0.415791   -0.077909   -0.043989    0.103141  ...   -0.023028   
3   -0.191507   -0.125910    0.200092   -0.013143  ...   -0.004660   
4   -0.185925    0.564402   -0.209667   -0.318550  ...   -0.003939   

   cg02298570  cg03545635  cg07310951  cg02712878  cg04697562  cg20090467  \
0   -0.042265  

Patient Number  cg22051636  cg25569462  cg03606258  cg22022041  cg00729708  \
0               1    0.028607    0.118231    0.354948   -0.195587   -0.328310   
1               2   -0.421796   -0.484327    0.358169    0.184083    0.058353   
2               3   -0.425853   -0.486174    0.017406   -0.183193    0.227653   
3               4   -0.449498    0.136772    0.366528   -0.169309    0.151502   
4               5    0.230452    0.232878    0.313735    0.187171   -0.371522   

   cg07747299  cg14159672  cg12437481  cg20945531  ...  cg25307081  \
0   -0.175936   -0.169272    0.166648    0.013308  ...   -0.043515   
1   -0.173480   -0.135551   -0.196084    0.024734  ...   -0.016707   
2    0.415791   -0.077909   -0.043989    0.103141  ...   -0.023028   
3   -0.191507   -0.125910    0.200092   -0.013143  ...   -0.004660   
4   -0.185925    0.564402   -0.209667   -0.318550  ...   -0.003939   

   cg02298570  cg03545635  cg07310951  cg02712878  cg04697562  cg20090467  \
0   -0.042265   -0.087791   -0.030244   -0.002556   -0.013288   -0.018405   
1   -0.020371    0.002521   -0.022933   -0.042472   -0.011147   -0.048486   
2   -0.007988    0.001124    0.033117   -0.066167    0.030863   -0.007400   
3    0.007814    0.016429    0.007355   -0.021429    0.031454    0.004433   
4    0.007787    0.036709   -0.031628    0.018501    0.004436    0.007735   

   cg24150528  cg17036737        Age  
0    0.028344    0.041416  20.786301  
1   -0.000823   -0.098725  18.158904  
2    0.012948   -0.000146  57.868493  
3    0.026085    0.015497  19.706849  
4    0.001121   -0.015098  61.956164  

[5 rows x 1037 columns]

In [154]:
#Using any regression model, predict the age of the patients
#First we split the data into training and testing sets
X = df_sites.drop('Age', axis=1) #The Features
y = df_sites['Age'] #The Target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) #split, #training size is 80%


In [155]:
#We inisialise a Linear Regression model for feature selection
linear_model_fs = LinearRegression()

#Recursive Feature Elimination
#I select 1000 feature here
rfe = RFE(estimator=linear_model_fs, n_features_to_select=1000)
rfe.fit(X_train, y_train)

#Get and print the selected features
selected_features = X_train.columns[rfe.support_]
print("Selected Features:", selected_features)

Selected Features: Index(['cg22051636', 'cg25569462', 'cg03606258', 'cg22022041', 'cg00729708',
       'cg07747299', 'cg14159672', 'cg12437481', 'cg20945531', 'cg13284426',
       ...
       'cg14766682', 'cg21096966', 'cg02298570', 'cg03545635', 'cg07310951',
       'cg02712878', 'cg04697562', 'cg20090467', 'cg24150528', 'cg17036737'],
      dtype='object', length=1000)


In [156]:
#Perform linear regression
linear_model = LinearRegression()
linear_model.fit(X_train[selected_features], y_train)

#Predict on the test set
y_pred = linear_model.predict(X_test[selected_features])
print(y_pred)

#Calculate the mean squared error
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")

#Evaluate the model by calculating the R-squares (two ways of doing that, below),
#R-squared indicates the goodness of fit. The closer to 1, the better the model.
print(f"Model Score: {linear_model.score(X_test[selected_features], y_test)}")
#Calculate R-squared
r2 = r2_score(y_test, y_pred)
print(f"R-squared of the model with 1000 features: {r2}")

[-0.80372162 54.69630131 52.58331819  3.73511047 -2.83698005 24.56249175
 20.87867524 44.61591692 27.33455787 37.78493986 -2.28402732 -0.6175883
 25.2800651  -2.81837097 63.41609393 11.16821981 55.55604043  3.06656947
  3.77576867 50.15523363 62.23933925 48.61335877]
Mean Squared Error: 103.63116266542593
Model Score: 0.8503146365011985
R-squared of the model with 1000 features: 0.8503146365011985


In [157]:
#Perform the Random Forest Regression
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train[selected_features], y_train)

#Predict on the test set
y_pred_rf = rf_model.predict(X_test[selected_features])

#Calculate the mean squared error for Random Forest
mse_rf = mean_squared_error(y_test, y_pred_rf)

#print
print(f"Random Forest Regression MSE with feature selection: {mse_rf}")
#evaluate the model
print(f"Model Score: {rf_model.score(X_test[selected_features], y_test)}")


Random Forest Regression MSE with feature selection: 371.36086494911126
Model Score: 0.4636045314033478


By looking at the MSE and goodness of fit of the two models, the model with lower MSE and R2 is Linear regression. So, we will carry on using this model.

In [158]:
#Now we see the performance of the linear regression model while progressively reducing the features
#with 90 features
#We inisialise a Linear Regression model for feature selection
linear_model_fs = LinearRegression()

#Recursive Feature Elimination
#I select the first 90 feature here
rfe = RFE(estimator=linear_model_fs, n_features_to_select=90)
rfe.fit(X_train, y_train)

#Get and print the selected features
selected_features_90 = X_train.columns[rfe.support_]
print("Selected Features:", selected_features_90)

Selected Features: Index(['cg03606258', 'cg00729708', 'cg12437481', 'cg16106497', 'cg20507276',
       'cg18920397', 'cg01889448', 'cg20540428', 'cg05187322', 'cg23280807',
       'cg22432908', 'cg17711541', 'cg00298951', 'cg23183296', 'cg19668234',
       'cg12188860', 'cg15403517', 'cg14813083', 'cg21992250', 'cg05590257',
       'cg03882305', 'cg21116314', 'cg04987894', 'cg15842276', 'cg12454167',
       'cg14951292', 'cg03874127', 'cg08832227', 'cg24993443', 'cg26020513',
       'cg20903926', 'cg21561173', 'cg16112050', 'cg27644292', 'cg02946570',
       'cg21165219', 'cg00406188', 'cg02828104', 'cg21306775', 'cg18829411',
       'cg08952029', 'cg04622802', 'cg02729303', 'cg18540325', 'cg05590982',
       'cg10135717', 'cg09835085', 'cg00242839', 'cg17003970', 'cg22373112',
       'cg19107595', 'cg14592065', 'cg07997737', 'cg08961832', 'cg24378421',
       'cg19000186', 'cg02654291', 'cg06818473', 'cg24164563', 'cg04451770',
       'cg08459368', 'cg07109801', 'cg19428735', 'cg05982

In [159]:
#Perform linear regression
linear_model = LinearRegression()
linear_model.fit(X_train[selected_features_90], y_train)

#Predict on the test set
y_pred = linear_model.predict(X_test[selected_features_90])
print(f"selected features: {y_pred}")

#Calculate the mean squared error
mse_90 = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse_90}")

#Evaluate the model by calculating the R-squares (two ways of doing that, below),
#R-squared indicates the goodness of fit. The closer to 1, the better the model.
print(f"Model Score: {linear_model.score(X_test[selected_features_90], y_test)}")
#Calculate R-squared
r2 = r2_score(y_test, y_pred)
print(f"R-squared of the model with 90 features: {r2}")

selected features: [ -8.73782222  56.1336671   52.11187401  -6.33349641  -5.3527745
   8.64467945  29.3185273   44.24230135  34.54711149  37.05079612
 -14.47493548   8.8974723   28.25537565  -5.04647658  64.10295142
  -1.87829644  48.89354939   0.92942344  -0.41728901  55.67381843
  65.48483437  48.16091557]
Mean Squared Error: 114.73241479411664
Model Score: 0.8342799330641667
R-squared of the model with 90 features: 0.8342799330641667


In [160]:
#with 80 features
#We inisialise a Linear Regression model for feature selection
linear_model_fs = LinearRegression()

#Recursive Feature Elimination
#I select 80 feature here
rfe = RFE(estimator=linear_model_fs, n_features_to_select=80)
rfe.fit(X_train, y_train)

#Get and print the selected features
selected_features_80 = X_train.columns[rfe.support_]
print("Selected Features:", selected_features_80)

Selected Features: Index(['cg03606258', 'cg00729708', 'cg12437481', 'cg16106497', 'cg20507276',
       'cg18920397', 'cg01889448', 'cg20540428', 'cg05187322', 'cg23280807',
       'cg22432908', 'cg17711541', 'cg00298951', 'cg23183296', 'cg19668234',
       'cg12188860', 'cg15403517', 'cg21992250', 'cg05590257', 'cg03882305',
       'cg21116314', 'cg04987894', 'cg15842276', 'cg12454167', 'cg14951292',
       'cg03874127', 'cg08832227', 'cg24993443', 'cg26020513', 'cg20903926',
       'cg21561173', 'cg16112050', 'cg27644292', 'cg02946570', 'cg21165219',
       'cg00406188', 'cg02828104', 'cg21306775', 'cg18829411', 'cg08952029',
       'cg04622802', 'cg02729303', 'cg05590982', 'cg10135717', 'cg09835085',
       'cg00242839', 'cg17003970', 'cg22373112', 'cg19107595', 'cg07997737',
       'cg24378421', 'cg19000186', 'cg02654291', 'cg06818473', 'cg24164563',
       'cg04451770', 'cg07109801', 'cg19428735', 'cg05982504', 'cg14463304',
       'cg15407570', 'cg26206598', 'cg01680762', 'cg00333

In [161]:
#Perform linear regression with 80 features
linear_model = LinearRegression()
linear_model.fit(X_train[selected_features_80], y_train)

#Predict on the test set
y_pred = linear_model.predict(X_test[selected_features_80])
print(f"selected features: {y_pred}")

#Calculate the mean squared error
mse_80 = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse_80}")

#Evaluate the model by calculating the R-squares (two ways of doing that, below),
#R-squared indicates the goodness of fit. The closer to 1, the better the model.
print(f"Model Score: {linear_model.score(X_test[selected_features_80], y_test)}")
#Calculate R-squared
r2 = r2_score(y_test, y_pred)
print(f"R-squared of the model with 80 features: {r2}")

selected features: [-11.24356246  56.97507499  49.19166834  -7.61212576  -4.95035183
   9.62165437  28.27561782  45.86070452  31.26855812  33.96419536
 -11.82236963   5.30029077  27.51110028  -5.76505268  69.30267491
  -4.15760347  49.2860271    2.60843082  -1.70476795  53.84136342
  64.15134852  47.528615  ]
Mean Squared Error: 99.10555363012426
Model Score: 0.8568514485573331
R-squared of the model with 80 features: 0.8568514485573331


In [162]:
#with 70 features
#We inisialise a Linear Regression model for feature selection
linear_model_fs = LinearRegression()

#Recursive Feature Elimination
#I select 70 feature here
rfe = RFE(estimator=linear_model_fs, n_features_to_select=70)
rfe.fit(X_train, y_train)

#Get and print the selected features
selected_features_70 = X_train.columns[rfe.support_]
print("Selected Features:", selected_features_70)

Selected Features: Index(['cg03606258', 'cg00729708', 'cg12437481', 'cg20507276', 'cg18920397',
       'cg01889448', 'cg20540428', 'cg05187322', 'cg22432908', 'cg17711541',
       'cg00298951', 'cg23183296', 'cg19668234', 'cg12188860', 'cg15403517',
       'cg21992250', 'cg05590257', 'cg03882305', 'cg21116314', 'cg04987894',
       'cg15842276', 'cg12454167', 'cg14951292', 'cg03874127', 'cg08832227',
       'cg24993443', 'cg26020513', 'cg21561173', 'cg02946570', 'cg21165219',
       'cg00406188', 'cg02828104', 'cg21306775', 'cg18829411', 'cg08952029',
       'cg04622802', 'cg02729303', 'cg05590982', 'cg10135717', 'cg09835085',
       'cg00242839', 'cg17003970', 'cg22373112', 'cg19107595', 'cg07997737',
       'cg24378421', 'cg02654291', 'cg06818473', 'cg24164563', 'cg07109801',
       'cg05982504', 'cg14463304', 'cg26206598', 'cg01680762', 'cg00333528',
       'cg19636861', 'cg16708981', 'cg03918304', 'cg01457622', 'cg14223444',
       'cg21842274', 'cg19335327', 'cg23061257', 'cg06914

In [163]:
#Perform linear regression with 70 features
linear_model = LinearRegression()
linear_model.fit(X_train[selected_features_70], y_train)

#Predict on the test set
y_pred = linear_model.predict(X_test[selected_features_70])
print(f"selected features: {y_pred}")

#Calculate the mean squared error
mse_70 = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse_70}")

#Evaluate the model by calculating the R-squares (two ways of doing that, below),
#R-squared indicates the goodness of fit. The closer to 1, the better the model.
print(f"Model Score: {linear_model.score(X_test[selected_features_70], y_test)}")
#Calculate R-squared
r2 = r2_score(y_test, y_pred)
print(f"R-squared of the model with 70 features: {r2}")

selected features: [-1.53693117e+01  5.61036214e+01  5.16950644e+01 -7.90292662e+00
  5.33364831e-02  1.30753804e+01  3.20697505e+01  4.62304721e+01
  3.03508210e+01  3.54378199e+01 -1.40069269e+01  2.30342020e+00
  2.28960558e+01 -4.94919082e+00  7.46903884e+01 -5.12991462e-02
  5.36654584e+01  5.24868561e-01 -1.08024287e+01  5.72155315e+01
  6.27870752e+01  4.88857416e+01]
Mean Squared Error: 112.6604256334505
Model Score: 0.8372727244475964
R-squared of the model with 70 features: 0.8372727244475964


In [164]:
#with 60 features
#We inisialise a Linear Regression model for feature selection
linear_model_fs = LinearRegression()

#Recursive Feature Elimination
#I select 60 feature here
rfe = RFE(estimator=linear_model_fs, n_features_to_select=60)
rfe.fit(X_train, y_train)

#Get and print the selected features
selected_features_60 = X_train.columns[rfe.support_]
print("Selected Features:", selected_features_60)

Selected Features: Index(['cg00729708', 'cg12437481', 'cg20507276', 'cg20540428', 'cg05187322',
       'cg22432908', 'cg17711541', 'cg00298951', 'cg23183296', 'cg19668234',
       'cg12188860', 'cg15403517', 'cg21992250', 'cg05590257', 'cg03882305',
       'cg21116314', 'cg04987894', 'cg15842276', 'cg14951292', 'cg03874127',
       'cg08832227', 'cg24993443', 'cg26020513', 'cg21561173', 'cg02946570',
       'cg02828104', 'cg21306775', 'cg18829411', 'cg08952029', 'cg04622802',
       'cg02729303', 'cg05590982', 'cg10135717', 'cg09835085', 'cg00242839',
       'cg17003970', 'cg22373112', 'cg19107595', 'cg07997737', 'cg24378421',
       'cg02654291', 'cg06818473', 'cg07109801', 'cg05982504', 'cg14463304',
       'cg26206598', 'cg01680762', 'cg19636861', 'cg16708981', 'cg03918304',
       'cg14223444', 'cg19335327', 'cg23061257', 'cg06914598', 'cg26033681',
       'cg01667702', 'cg17786776', 'cg19202384', 'cg20582779', 'cg19372178'],
      dtype='object')


In [165]:
#Perform linear regression with 60 features
linear_model = LinearRegression()
linear_model.fit(X_train[selected_features_60], y_train)

#Predict on the test set
y_pred = linear_model.predict(X_test[selected_features_60])
print(f"selected features: {y_pred}")

#Calculate the mean squared error
mse_60 = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse_60}")

#Evaluate the model by calculating the R-squares (two ways of doing that, below),
#R-squared indicates the goodness of fit. The closer to 1, the better the model.
print(f"Model Score: {linear_model.score(X_test[selected_features_60], y_test)}")
#Calculate R-squared
r2 = r2_score(y_test, y_pred)
print(f"R-squared of the model with 60 features: {r2}")

selected features: [-22.02018165  53.63423872  49.53181439  -7.16359741   1.89623256
  14.23551532  27.94983311  44.20384661  43.29491663  33.06970847
 -10.70309155  10.98848573  27.39111201  -6.1801712   76.14298965
  -0.30607325  50.39808466  14.38099008  -9.58580786  55.25770737
  59.85531904  48.03875435]
Mean Squared Error: 177.32992626987183
Model Score: 0.743863778131893
R-squared of the model with 60 features: 0.743863778131893


In [166]:
#with 50 features
#We inisialise a Linear Regression model for feature selection
linear_model_fs = LinearRegression()

#Recursive Feature Elimination
#I select 50 feature here
rfe = RFE(estimator=linear_model_fs, n_features_to_select=50)
rfe.fit(X_train, y_train)

#Get and print the selected features
selected_features_50 = X_train.columns[rfe.support_]
print("Selected Features:", selected_features_50)

Selected Features: Index(['cg12437481', 'cg20540428', 'cg22432908', 'cg17711541', 'cg00298951',
       'cg23183296', 'cg19668234', 'cg12188860', 'cg15403517', 'cg21992250',
       'cg05590257', 'cg03882305', 'cg21116314', 'cg04987894', 'cg15842276',
       'cg14951292', 'cg03874127', 'cg08832227', 'cg26020513', 'cg02946570',
       'cg02828104', 'cg08952029', 'cg02729303', 'cg05590982', 'cg09835085',
       'cg00242839', 'cg17003970', 'cg22373112', 'cg19107595', 'cg07997737',
       'cg24378421', 'cg02654291', 'cg06818473', 'cg07109801', 'cg05982504',
       'cg14463304', 'cg26206598', 'cg01680762', 'cg19636861', 'cg16708981',
       'cg03918304', 'cg19335327', 'cg23061257', 'cg06914598', 'cg26033681',
       'cg01667702', 'cg17786776', 'cg19202384', 'cg20582779', 'cg19372178'],
      dtype='object')


In [167]:
#Perform linear regression with 50 features
linear_model = LinearRegression()
linear_model.fit(X_train[selected_features_50], y_train)

#Predict on the test set
y_pred = linear_model.predict(X_test[selected_features_50])
print(f"selected features: {y_pred}")

#Calculate the mean squared error
mse_50 = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse_50}")

#Evaluate the model by calculating the R-squares (two ways of doing that, below),
#R-squared indicates the goodness of fit. The closer to 1, the better the model.
print(f"Model Score: {linear_model.score(X_test[selected_features_50], y_test)}")
#Calculate R-squared
r2 = r2_score(y_test, y_pred)
print(f"R-squared of the model with 50 features: {r2}")

selected features: [-23.14488205  53.47154417  53.58781293 -11.62116668   4.48102825
  17.56519472  30.11657863  43.47531794  51.41617113  32.4855453
  -8.12933015   9.2363631   30.21162412  -7.72954351  81.07505394
  -3.75475806  57.72066961  13.09583533 -16.36553574  55.23984942
  58.47588001  50.33989529]
Mean Squared Error: 233.7925401024686
Model Score: 0.6623088996740409
R-squared of the model with 50 features: 0.6623088996740409


In [168]:
#with 40 features
#We inisialise a Linear Regression model for feature selection
linear_model_fs = LinearRegression()

#Recursive Feature Elimination
#I select 40 feature here
rfe = RFE(estimator=linear_model_fs, n_features_to_select=40)
rfe.fit(X_train, y_train)

#Get and print the selected features
selected_features_40 = X_train.columns[rfe.support_]
print("Selected Features:", selected_features_40)

Selected Features: Index(['cg12437481', 'cg17711541', 'cg00298951', 'cg23183296', 'cg19668234',
       'cg12188860', 'cg21992250', 'cg05590257', 'cg03882305', 'cg21116314',
       'cg04987894', 'cg14951292', 'cg03874127', 'cg26020513', 'cg02946570',
       'cg02828104', 'cg08952029', 'cg09835085', 'cg17003970', 'cg22373112',
       'cg19107595', 'cg07997737', 'cg24378421', 'cg02654291', 'cg06818473',
       'cg07109801', 'cg26206598', 'cg01680762', 'cg19636861', 'cg16708981',
       'cg03918304', 'cg19335327', 'cg23061257', 'cg06914598', 'cg26033681',
       'cg01667702', 'cg17786776', 'cg19202384', 'cg20582779', 'cg19372178'],
      dtype='object')


In [169]:
#Perform linear regression with 40 features
linear_model = LinearRegression()
linear_model.fit(X_train[selected_features_40], y_train)

#Predict on the test set
y_pred = linear_model.predict(X_test[selected_features_40])
print(f"selected features: {y_pred}")

#Calculate the mean squared error
mse_40 = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse_40}")

#Evaluate the model by calculating the R-squares (two ways of doing that, below),
#R-squared indicates the goodness of fit. The closer to 1, the better the model.
print(f"Model Score: {linear_model.score(X_test[selected_features_40], y_test)}")
#Calculate R-squared
r2 = r2_score(y_test, y_pred)
print(f"R-squared of the model with 40 features: {r2}")

selected features: [-24.70842389  50.1687061   52.36906475 -18.95670047   0.430189
  12.38417945  25.51929087  48.3240617   62.80141512  39.67489893
  -9.11685337   2.43361024  25.79211918   1.12235957  78.57080527
  -5.32300682  62.46408677   6.55305064 -18.1270585   57.50350311
  46.14625989  51.11659989]
Mean Squared Error: 295.16774613329596
Model Score: 0.5736582487670505
R-squared of the model with 40 features: 0.5736582487670505


In [170]:
#with 30 features
#We inisialise a Linear Regression model for feature selection
linear_model_fs = LinearRegression()

#Recursive Feature Elimination
#I select 30 feature here
rfe = RFE(estimator=linear_model_fs, n_features_to_select=30)
rfe.fit(X_train, y_train)

#Get and print the selected features
selected_features_30 = X_train.columns[rfe.support_]
print("Selected Features:", selected_features_30)

Selected Features: Index(['cg00298951', 'cg23183296', 'cg21992250', 'cg05590257', 'cg04987894',
       'cg14951292', 'cg03874127', 'cg26020513', 'cg02946570', 'cg02828104',
       'cg08952029', 'cg09835085', 'cg17003970', 'cg22373112', 'cg19107595',
       'cg07997737', 'cg02654291', 'cg06818473', 'cg07109801', 'cg01680762',
       'cg19636861', 'cg16708981', 'cg19335327', 'cg23061257', 'cg26033681',
       'cg01667702', 'cg17786776', 'cg19202384', 'cg20582779', 'cg19372178'],
      dtype='object')


In [171]:
#Perform linear regression with 30 features
linear_model = LinearRegression()
linear_model.fit(X_train[selected_features_30], y_train)

#Predict on the test set
y_pred = linear_model.predict(X_test[selected_features_30])
print(f"selected features: {y_pred}")

#Calculate the mean squared error
mse_30 = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse_30}")

#Evaluate the model by calculating the R-squares (two ways of doing that, below),
#R-squared indicates the goodness of fit. The closer to 1, the better the model.
print(f"Model Score: {linear_model.score(X_test[selected_features_30], y_test)}")
#Calculate R-squared
r2 = r2_score(y_test, y_pred)
print(f"R-squared of the model with 30 features: {r2}")

selected features: [-18.06768915  54.88646354  42.0473102  -22.41022276   2.47223195
   8.5557027   15.32875402  51.15796677  67.24759534  51.00880243
 -14.99368885  -2.48028967  26.48143874  12.45951991  85.3250543
   0.33032097  62.8495432   -1.21660515 -23.84708395  58.59443158
  41.49946817  57.98773153]
Mean Squared Error: 375.03703999452114
Model Score: 0.45829464600011294
R-squared of the model with 30 features: 0.45829464600011294


In [172]:
#with 20 features
#We inisialise a Linear Regression model for feature selection
linear_model_fs = LinearRegression()

#Recursive Feature Elimination
#I select 20 feature here
rfe = RFE(estimator=linear_model_fs, n_features_to_select=20)
rfe.fit(X_train, y_train)

#Get and print the selected features
selected_features_20 = X_train.columns[rfe.support_]
print("Selected Features:", selected_features_20)

Selected Features: Index(['cg21992250', 'cg05590257', 'cg04987894', 'cg03874127', 'cg02828104',
       'cg08952029', 'cg09835085', 'cg17003970', 'cg19107595', 'cg07997737',
       'cg02654291', 'cg06818473', 'cg07109801', 'cg19636861', 'cg16708981',
       'cg23061257', 'cg01667702', 'cg17786776', 'cg20582779', 'cg19372178'],
      dtype='object')


In [173]:
#Perform linear regression with 20 features
linear_model = LinearRegression()
linear_model.fit(X_train[selected_features_20], y_train)

#Predict on the test set
y_pred = linear_model.predict(X_test[selected_features_20])
print(f"selected features: {y_pred}")

#Calculate the mean squared error
mse_20 = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse_20}")

#Evaluate the model by calculating the R-squares (two ways of doing that, below),
#R-squared indicates the goodness of fit. The closer to 1, the better the model.
print(f"Model Score: {linear_model.score(X_test[selected_features_20], y_test)}")
#Calculate R-squared
r2 = r2_score(y_test, y_pred)
print(f"R-squared of the model with 20 features: {r2}")

selected features: [-15.61926153  67.13670493  31.95746977 -12.84046626  -3.11406082
   3.41275494   7.76029686  52.56814685  79.81115484  48.59170903
 -19.46465778  -7.79126875  11.79961149  16.06689798  67.91412296
  13.32769553  61.96695663 -12.78891151  -7.45979386  59.00231646
  44.12795056  57.6302427 ]
Mean Squared Error: 451.8787121429287
Model Score: 0.34730415499766554
R-squared of the model with 20 features: 0.34730415499766554


In [174]:
#with 10 features
#We inisialise a Linear Regression model for feature selection
linear_model_fs = LinearRegression()

#Recursive Feature Elimination
#I select 10 feature here
rfe = RFE(estimator=linear_model_fs, n_features_to_select=10)
rfe.fit(X_train, y_train)

#Get and print the selected features
selected_features_10 = X_train.columns[rfe.support_]
print("Selected Features:", selected_features_10)

Selected Features: Index(['cg21992250', 'cg03874127', 'cg02828104', 'cg09835085', 'cg07997737',
       'cg02654291', 'cg16708981', 'cg23061257', 'cg01667702', 'cg20582779'],
      dtype='object')


In [175]:
#Perform linear regression with 10 features
linear_model = LinearRegression()
linear_model.fit(X_train[selected_features_10], y_train)

#Predict on the test set
y_pred = linear_model.predict(X_test[selected_features_10])
print(f"selected features: {y_pred}")

#Calculate the mean squared error
mse_10 = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse_10}")

#Evaluate the model by calculating the R-squares (two ways of doing that, below),
#R-squared indicates the goodness of fit. The closer to 1, the better the model.
print(f"Model Score: {linear_model.score(X_test[selected_features_10], y_test)}")
#Calculate R-squared
r2 = r2_score(y_test, y_pred)
print(f"R-squared of the model with 10 features: {r2}")

selected features: [-20.79986522  47.19589886  34.63533323 -12.10166253 -14.03780821
  -5.39410746  18.51781126  57.11789569  80.80783803  52.80810295
 -14.72714905  -2.98912268   9.70415454  19.55991677  63.60317054
   0.09916861  70.08026549 -22.39914218   3.63851844  52.96614619
  55.76741363  61.6404005 ]
Mean Squared Error: 488.04359708618034
Model Score: 0.2950674164588921
R-squared of the model with 10 features: 0.2950674164588921


**Considerations:**
Jumping from 1000 straight to 90 feature does not seem to affect the model significantly as the Rsquared value is still 0.85 (similarly for MSE values). Also, from 90 to 70 features the Rsquared (and the MSE) is still not significantly changing. However, by progressively decreasing the feature from 70 to 10, there is a decrease in the Rsquared and an increase of the corrispecting MSE, with a low of 0.29 and 488, with only 10 feature. Therefore, it seems that keeping 90/80/70 features might be the best choice in this case, as we want to keep the MSE, which is the mean square error, as low as possible and the RSquare as closest to 1 as possible.

In [176]:
#Generate a plot of the number of features against the MSE value
#to do that, first I need to create a new dataset with the specific data
# We Create a dictionary to store the data of interest
data = {
    'Number of Features': [90, 80, 70, 60, 50, 40, 30, 20, 10],

    'MSE': [mse_90, mse_80, mse_70, mse_60, mse_50, mse_40, mse_30, mse_20, mse_10]
}

# Create a Pandas DataFrame from the dictionary
feature_mse_df = pd.DataFrame(data)

# Print the DataFrame
print(feature_mse_df)


   Number of Features         MSE
0                  90  114.732415
1                  80   99.105554
2                  70  112.660426
3                  60  177.329926
4                  50  233.792540
5                  40  295.167746
6                  30  375.037040
7                  20  451.878712
8                  10  488.043597


In [180]:
# This is the code for plotting the features against the MSE, which an indication of the performance of the model
#We are choosing to represent the data with a scatter plot

import plotly.express as px

fig = px.scatter(feature_mse_df,
             x='Number of Features',
             y='MSE',
             color_discrete_sequence=['cyan'])  # Set color to cyan

fig.update_traces(marker_line_color='blue', marker_line_width=1.5, marker=dict(size=10))  # Add border to bars

fig.update_layout(
    xaxis_title="Number of Features",  # Add x-axis label
    yaxis_title="MSE Values",  # Add y-axis label
    xaxis=dict(type='category')  # Set x-axis type to category

)

fig.show()